### ISSUE TITLE
Some way to define hash-tag like things in the content

### ISSUE THREAD

**Issue by jkomoros** [jkomoros/card-web#399]
E.g. other knowledge gardens allow putting double-brackets around words that should be treated as though they're a hashtag or tag, and not just random text. This allows the card author to communicate that it's not just arbitrary that these words occur next to each other, but that they mean something specific.

And then ideally these tags would all be normalized and look for overlap, etc so they could be misspelled.

Related: bold text should be treated as though it shows up like 5x as often in the text of the card, so it's very, very distinctive

**Comment by jkomoros** [comment]
Perhaps these special-emphasis hashtags would show up as a special highlighted color in word clouds.

Ideally doing this would be fast to do in content-editable, and the resulting HTML would be valid HTML.

Perhaps have a thing like card-link that's kind of like the special `strong` tag? That would have a `value` attribute, which if not set would default to the text inside the tag. (So the display information could be different than the 'semantic' value). And each of these could be rendered specially and clickable to find other cards that have the same tag.

When a card is saved, its HTML fields would be run through and have any of these tags extracted out of them and indexed separately. Words that are in these specia lruns of text would have a very high match score, and would also be treated as though they showed up like 5x in the content of the card.

These hashtags would kind of be like normal tags, but they'd be fully implied by content.

**Comment by jkomoros** [comment]
Maybe call these `concept-tag`. And then there would be tools to find related or overlapping concepts, to help find overlapping or mispelled ones. (Concept tags would have fingerprints be calculated similar to tags). 

There would be a filter, `concept/sirens-call` that returns all of the cards that have that concept tag.

When editing, there could be a feature where related concept tags are suggested. (Can you include a concept tag without it literally showing up in the prose?)

The actual concept tag canonical representation would be the stemmed, normalized run of the concept value

**Comment by jkomoros** [comment]
Some kind of feature to find cards that seem to have the concept tag (e.g. they have the n-gram in their fingerprint, but don't yet point to the concept tag directly)

**Comment by jkomoros** [comment]
Worth thinking about: should concept tags be modeled as references to a concept card? That fits with the everything-as-card model. It would need the machinery that allows editing meta information as though it's implicitly part of the card, because the concept card would primarily be just a collection of links about that concept. 

In general, there could be a `<reference-block collection='descendants/SELF/2'></reference-block>` that can be put into a card. That item automatically expands when the card is viewed to render links for each card that matches the collection. `SELF` is a special stand-in that will be replaced by this card's ID before the collection is fetched. As far as indexing and editing is concerned, that block of references doesn't exist. 

Actually, these reference blocks shouldn't be literal items in the HTML, but should be extra metadata on the card:

```
{
  //...
  reference_blocks: {
    'similar/SELF/limit/10/': 'Similar',
    'children': 'Children',
    'query/foo/': '',
  }
  //...
}
```

Where the text SELF is matched and replaced with the card's own ID. The values in the dict are the display names for that collection, '' is legal and means 'don't describe what they are'. There are various aliases predefined, e.g. `children` expands to `children/SELF/` and `similar` expands to `similar/SELF/limit/10`. If the display name text is empty, and the alias has pre-defined display name, it will use that instead.

And then the card's main content gets flex 1, and the reference blocks get the remaining space, scrolling if necessary to fit.

**Comment by jkomoros** [comment]
If we do the reference-blocks as described immediately above, then `<concept-tag>` is effectively a card-link with a reference type of concept. 

And concept cards don't get any todos about needed reciprocal links.

**Comment by jkomoros** [comment]
This comment tries to distill how all of this works based on the exploration comments above.

There's a new type of card, `concept`. These cards denote a 'concept' that shows up in other cards, and this is the canonical location they're described. They have a title which is the canonical name of the concept, a body which is the definition of them, and then they have a reference_block for the card type (see #400) that is defined as `references-direct/SELF/inbound/concept/`, so those inbound references show up in the card. Cards of this type are not part of the default set, but are part of everything. Cards of this type warn if they are saved but not published.



concept cards created via linking from find dialog (created concept cards start out published)

**Comment by jkomoros** [comment]
Looking through how to do the backporting of reference text, and it occurs to me: why is slugIndex ever factored into state, and why is normalized text for card not done via a selector? Both could be calculated via selectors.

**Comment by jkomoros** [comment]
- [ ] Create base concept card
- [ ] Concept cards, and working-notes cards, show up in `everything/concept` canonically. 
- [ ] Concept cards get a little book icon at the from of title in card and thumbnail
- [ ] Pull in TODO items from mondo comment above

**Comment by jkomoros** [comment]
One of the hardest things to do will be the backporting of reference text.

So much of the machinery assumes that each card is independent and doesn't need information from other cards when they load (with the exception of references, which are computed once on the server).

Backporting of reference text will be _relatively_ rare. The amount of overhead should be linear or less with the amount of backporting being done in the corpus, and it shouldn't have to be recalculated any time _any_ cards change, but ideally only when relied-upon cards change. And the machinery will get weirder when reference-type `synonym` comes into being, which might have a whole clique of synonyms that have to be backported across all of them.

One option is to extend the updateInboundLinks machinery to do the backporting then, and save it in the data model. **TODO** explore that option design.

Another option is to do it clientside. We'd take advantage of the fact that inbound references have enough information so that every card that is being relied on knows it is being relied on.

Have a `selectReferenceInfoForBackPorting(cards)` that returns a map of `fromCardID -> toCardID -> referenceType -> information`, which is selected based on filtering down to cards that have an inbound reference of a backportable type, with no text configured on the inbound reference, meaning it will need to be backported. It would use reselect-map inside, so the output wouldn't have to change if none of the affected cards changed.

There could be a notion of a referencesOverlay, and the selector named in the above paragraph is a referencesOverlay that is overlaid over the given card for purposes of nlp.

Another question is: is the normalized/extracted text info that the card uses actually stored in the datamodel, or is it stored in reselected selectors? If it were done in optimized selectors that used reselectMap, and depended on a conservatively-changing selectReferenceOverlayForBackporting(cards), then it wouldn't have to be that weird. In fact, we could have this all downstream of updateCards, so the normalized text search properies are munged in before upstream things deal with them in the first place. The reselect-map behavior will be key to make sure we don't constantly recalculate all of this stuff...

The one oddity of this approach is EDITING_TEXT_FIELD_UPDATED, where we want some way to delay the indexing/fingerprinting of the editing card to a one-second timeout so typing isn't a pain. When it was literally in the data model in state, that was easy--just setTimeout an action creator that asks for the indexing to happen. But it's a little bit harder if it's all implied via selectors, because you'd need to just delay _asking_ for it for a second. Maybe just have the info panel, if editing is active, have a setTimeout in the stateChanged() that delays asking for the reference blocks?

**Comment by jkomoros** [comment]
The wordcloud fingerprint for the same card when it's editing and when it's a normal card is slightly different, and it's not clear why. (It likely has to do with the card I'm testing with having such distinctive words that they can change the distribution if they're not part of the original set literally... but it also might be too-aggressive caching of selectCards r something)

Another issue: a longstanding bug in _cardTFIDF where it multiplies by this._idfMap[word]. But there could be words that are so rare that they don't show up in the rset of the corpus. That can't happen with normal fingerprints (they were inherently part of the corpus0, but that can happen with fingerprintForCardObj, where it might not have any items. The answer is to default to a 'very distinctive' tfidf if one doesn't exist in the corpus yet, preventing the NaN.

**Comment by jkomoros** [comment]
Regarding how to automatically have new concept cards have their name set to their (normalized) title. The key problem is that slugLegal is a cloud function, and it can take a really, really long time. We don't let a card be saved if it relies on a card that isn't yet fully created (since it could fail), so that delay would be a pain in the butt for it to check if slugLegal is true.

We could do a thing where we create the card, and then when it's loaded issue a modifyCard to set the slug. That way the card is created, everything else is unblocked, and it's just that we issue a new modifyCard at a later time. 

The other approach is to have a cloud function that looks for new cards, and an autoSlug: proposedSlug name, which if legal it sets as a slug. This could happen immediately (the cloud function can do the check if the slug is legal immediately) but at the cost of introducing a lot of bespoke machinery in cloud functions for what seems like a pretty specialized thing. 

The former is probably better. So at the end of modifyCard, if the cardType has a autoSlug: true in its config, then after the editing is done successfully, then it awaits the card to be loaded, then does a modifyCard. It can kick off the slugLegal check before commiting the new card to the db, so it's running concurrently, and only await its result later.
